Python/Module6_Testing/Intro_to_Testing.md

```diff
@@ -0,0 +1,532 @@
+
+Admittedly, the `count_vowels` function is simple enough that the inclusion of this assertion is rather pedantic.
+That being said, as we write increasingly sophisticated code, we will find that this sort of assertion will help us catch bad internal logic and oversights within our code base.
+We will also see that keen use of assertions can make it much easier for us to write good tests.
+<!-- #endregion -->
+
+## Testing Our Tests
+
+It is surprisingly easy to unwittingly write a broken test: a test that always passes, or a test that simply doesn't exercise our code in the way that we had intended.
+Broken tests are insidious; they are alarms that fail to sound when they are supposed to.
+They create misdirection in the bug-finding process and can mask problems with our code.
+**Thus a critical step in the test-writing process is to intentionally mutate the function of interest - to corrupt its behavior so that we can verify that our test works.**
+Once we confirm that our test does indeed raise an error as-expected, we restore the function to its original form and re-run the test and see that it passes. 
+
+A practical note: we ought to mutate our function in a way that is trivial to undo. We can use of code-comments towards this end.
+All [IDEs](https://www.pythonlikeyoumeanit.com/Module1_GettingStartedWithPython/Getting_Started_With_IDEs_and_Notebooks.html) have the ability to "block-comment" selected code.
+In order to block-comment code in a Jupyter notebook code cell, highlight the lines of code and press `CTRL + /`.
+The same key-combination will also un-comment a highlighted block of commented code.
+
+
+
+<div class="alert alert-info"> 
+
+**Reading Comprehension: Testing Your Test via Manual Mutation**
+
+Temporarily change the body of `count_vowels` such that the second assertion in `test_count_vowels_basic` raises an error.
+Run the test to confirm that the second assertion raises,
+and then restore `count_vowels` to its original form.
+Finally, rerun the test to see that `count_vowels` once again passes all of the assertions.
+    
+Repeat this process given the test that you wrote for `merge_max_mappings`.
+Try breaking the function such that it always merges in values from `dict2`, even if those values are smaller.
+
+</div>
+
+
+
+## Our Work, Cut Out
+
+We see now that the concept of a "test function" isn't all that fancy.
+Compared to other code that we have written, writing a function that simply runs a hand full of assertions is far from a heavy lift for us.
+Of course, we must be diligent and take care to test our tests, but we can certainly manage this as well.
+With this in hand, we should take stock of the work and challenges that lie in our path ahead.
+
+It is necessary that we evolve beyond manual testing.
+There are multiple facets to this observation.
+First, we must learn how to organize our test functions into a test suite that can be run in one fowl swoop.
```

Comment:
["fell swoop"](https://www.phrases.org.uk/meanings/at-one-fell-swoop.html)

Reply:
This is the second time I have done this 🙉 !